commons-geometry-examples/examples-jmh/src/main/java/org/apache/commons/geometry/examples/jmh/euclidean/pointmap/RebuildingKDTree.java

```diff
@@ -0,0 +1,259 @@
+            return findMedianStart(nodes, startIdx, endIdx, cutDimension);
+        }
+    }
+
+    /** Find the starting index of the node median value in the given list. The list is
+     * partially sorted and value less than the median come before the returned index while
+     * values greater than or equal come after.
+     * @param nodes list of node
+     * @param startIdx sublist start index (inclusive)
+     * @param endIdx sublist end index (exclusive)
+     * @param cutDimension cut dimension
+     * @return index of the median in the specific sublist of {@code nodes} along the cut dimension
+     */
+    protected int findMedianStart(final List<KDTreeNode<V>> nodes, final int startIdx, final int endIdx,
+            final CutDimension cutDimension) {
+        int k = startIdx + ((endIdx - startIdx) / 2);
+        int low = startIdx;
+        int high = endIdx - 1;
+        int lowTemp;
+        int highTemp;
+        double x;
+        while (low < high) {
+            x = cutDimension.getCoordinate(nodes.get(k).getKey());
+            lowTemp = low;
+            highTemp = high;
+            do {
+                while (cutDimension.getCoordinate(nodes.get(lowTemp).getKey()) < x) {
+                    ++lowTemp;
+                }
+                while (cutDimension.getCoordinate(nodes.get(highTemp).getKey()) > x) {
+                    --highTemp;
+                }
+
+                if (lowTemp <= highTemp) {
+                    swap(nodes, lowTemp, highTemp);
+
+                    ++lowTemp;
+                    --highTemp;
+                }
+            } while (lowTemp <= highTemp);
+
+            if (k < lowTemp) {
+                // search low part
+                high = highTemp;
+            }
+            if (k > highTemp) {
+                // search high part
+                low = lowTemp;
+            }
+        }
+
+        // back up to the start of the median value
+        x = cutDimension.getCoordinate(nodes.get(k).getKey());
+        while (k > startIdx &&
+                cutDimension.getCoordinate(nodes.get(k - 1).getKey()) == x) {
+            --k;
+        }
+
+        return k;
+    }
+
+    /** Construct a comparator the sorts by the given cut dimension.
+     * @param cutDimension cut dimension to sort by
+     * @return comparator along the cut dimension
+     */
+    protected Comparator<KDTreeNode<V>> comparator(final CutDimension cutDimension) {
+        return (a, b) -> Double.compare(cutDimension.getCoordinate(a.getKey()), cutDimension.getCoordinate(b.getKey()));
+    }
+
+    /** Swap two elements in {@code list}.
+     * @param <T> List element type
+     * @param list input list
+     * @param aIdx index of the first element
+     * @param bIdx index of the second element
+     */
+    private static <T> void swap(final List<T> list, final int aIdx, final int bIdx) {
+        final T temp = list.get(aIdx);
+        list.set(aIdx, list.get(bIdx));
+        list.set(bIdx, temp);
```

Comment:
Couldn't we use `Collections.swap(list, aIdx, bIdx);` here?

Reply:
I didn't even realize that was a thing! Thanks!